### DISCUSSION TITLE
Window remains on screen after being called from Godot as dynamic library.

### DISCUSSION THREAD

**Discussion by juliohq** [fltk-rs/fltk-rs#1363] [Q&A] [answered]
![preview](https://user-images.githubusercontent.com/33421921/221416368-7731cf02-fc97-4b7a-8eff-59a4f3bbf77d.gif)
I was doing some experiments with Godot Engine GDNative bindings for Rust (https://godot-rust.github.io) and for some reason, after closing the window, it freezes and widgets don't respond anymore while the thread is unblocked and I can continue to use the Godot window. Anyone tried compiling a fltk-rs application as a dynamic library too?

fltk-rs window freezed:
![image](https://user-images.githubusercontent.com/33421921/221412464-72d330f0-c30c-4200-84da-4d22fe387fc4.png)

Example library code:

```rust
use fltk::{prelude::*, app::App, window::Window, button::Button};

#[no_mangle]
pub extern "C" fn show() {
    let app = App::default();
    let mut win = Window::default().with_size(400, 400);
    let mut but = Button::default_fill().with_label("Click here.");
    win.end();
    win.show();
    but.set_callback(move |_| app.quit());
    app.run();
}
```

Example binary:

```rust
fn main() {
    call_dynamic();
    
    loop {
        std::thread::sleep(std::time::Duration::from_secs(1));
    }
}

fn call_dynamic() -> Result<u32, Box<dyn std::error::Error>> {
    unsafe {
        let lib = libloading::Library::new("../target/release/libfltk_lib.so")?;
        let func: libloading::Symbol<unsafe extern fn() -> u32> = lib.get(b"show")?;
        Ok(func())
    }
}
```

**Comment by MoAlyousef** [comment, marked as answer]
I'm not sure what the issue might be.

I've just tried with the X11 bindings, it seems since the display is open, xlib doesn't unmap the fltk window since they're in the same process if the fltk event loop is stopped suddenly. 

I would suggest setting the callback to be something like:
```rust
but.set_callback(move |_| { win.hide(); app::flush(); });
```
This is the equivalent of calling XFlush on X11.

Not sure how applicable this is with Godot, but also not that xlib doesn't handle multiple UI threads well. 

Anyways this is the code I tested with:
```rust
use x11::xlib::*;

fn main() {
    let fltk_win = call_dynamic().unwrap();
    unsafe {
        let d = XOpenDisplay(std::ptr::null());
        if d.is_null() {
            panic!("Can't open display!");
        }
        let s = XDefaultScreen(d);
        let w = XCreateSimpleWindow(
            d,
            XRootWindow(d, s),
            10,
            10,
            100,
            100,
            1,
            XBlackPixel(d, s),
            XWhitePixel(d, s),
        );

        let msg = "MyWin\0";

        XSelectInput(d, w, ExposureMask | KeyPressMask);
        XMapWindow(d, w);

        let mut e: XEvent = std::mem::zeroed();
        loop {
            XNextEvent(d, &mut e);
            if e.type_ == Expose {
                XFillRectangle(d, w, XDefaultGC(d, s), 20, 20, 10, 10);
                XDrawString(
                    d,
                    w,
                    XDefaultGC(d, s),
                    10,
                    50,
                    msg.as_ptr() as _,
                    msg.len() as _,
                );
            }
            if e.type_ == KeyPress {
                break;
            }
        }
        XCloseDisplay(d);
    }
}

fn call_dynamic() -> Result<i32, Box<dyn std::error::Error>> {
    unsafe {
        let lib = libloading::Library::new("../winbg/target/debug/libwinbg.so")?;
        let func: libloading::Symbol<unsafe extern "C" fn() -> i32> = lib.get(b"show")?;
        Ok(func())
    }
}
```

**Reply by juliohq** [reply]
That doesn't happen only in Godot. In my test with just a small Rust binary that calls the dynamic library, it doesn't close too. Maybe that's related to something living for the whole binary lifetime?

**Reply by MoAlyousef** [reply]
Did you notice any change with:
```rust
but.set_callback(move |_| { win.hide(); app::flush(); });
```

**Reply by juliohq** [reply]
> Did you notice any change with:
> 
> ```rust
> but.set_callback(move |_| { win.hide(); app::flush(); });
> ```

Yes, that works! Even without the x11 part.

**Reply by juliohq** [reply]
This works too:
```rust
but.set_callback(move |_| {win.hide(); app.quit(); });
```
